goto_handler.go

```diff
@@ -0,0 +1,118 @@
+package migrate
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// Define a constant for the migration file name
+const lastSuccessfulMigrationFile = "lastSuccessfulMigration"
+
+func (m *Migrate) HandleDirtyState(path string) {
+	// Perform actions when dirty is true
+	lastSuccessfulMigrationPath := filepath.Join(path, lastSuccessfulMigrationFile)
+	lastVersionBytes, err := os.ReadFile(lastSuccessfulMigrationPath)
+	if err != nil {
+		log.Fatalf("failed to read last successful migration file: %s", err)
+	}
+	lastVersionStr := strings.TrimSpace(string(lastVersionBytes))
+	lastVersion, err := strconv.ParseUint(lastVersionStr, 10, 64)
+	if err != nil {
+		log.Fatalf("failed to parse last successful migration version: %s", err)
+	}
+
+	if err := m.Force(int(lastVersion)); err != nil {
+		log.Fatalf("failed to apply last successful migration: %s", err)
+	}
+
+	log.Println("Last successful migration applied")
+
+	if err := os.Remove(lastSuccessfulMigrationPath); err != nil {
+		log.Fatalf("failed to delete last successful migration file: %s", err)
+	}
+
+	log.Println("Last successful migration file deleted")
+}
+
+func (m *Migrate) HandleMigrationFailure(curVersion int, v uint, path string) error {
+	if err := m.lock(); err != nil {
+		return m.unlockErr(err)
+	}
+
+	ret := make(chan interface{}, m.PrefetchMigrations)
+	go m.read(curVersion, int(v), ret)
+
+	var migrations []int
+	for r := range ret {
+		migrations = append(migrations, int(r.(*Migration).Version))
+	}
+
+	failedVersion, _, err := m.databaseDrv.Version()
+	if err != nil {
+		return err
+	}
+	log.Println("failedVersion:", failedVersion, "migrations:", migrations, curVersion, v)
```

Comment:
nit: You might want to add a prefix messages for `curVersion` and `v` so it's easier to understand when reading the message.

Reply:
for my testing purposes, I added this will remove at the end

Reply:
Up to you. But if you keep any of the added `log.Printf()`, you need to enclose them with `if log.verbose {}` to follow the existing pattern.